modules/search-solr-client/src/main/scala/io/renku/search/solr/documents/Entity.scala

```diff
@@ -34,15 +46,20 @@ final case class Project(
     createdBy: users.Id,
     creationDate: projects.CreationDate,
     score: Option[Double] = None
-)
+) extends Entity
 
 object Project:
   val entityType: String = "Project"
-
   given Encoder[Project] = deriveWithDiscriminator
-  given Decoder[Seq[User]] =
-    Decoder[Seq[User]] { reader =>
-      if reader.hasArrayStart then Decoder.forArray[User].map(_.toSeq).read(reader)
-      else Decoder[User].map(Seq(_)).read(reader)
-    }
-  given Decoder[Project] = deriveDecoder
+
+final case class User(
+    id: users.Id,
+    firstName: Option[users.FirstName] = None,
+    lastName: Option[users.LastName] = None,
+    email: Option[users.Email] = None,
+    score: Option[Double] = None
+) extends Entity
+
+object User:
+  val entityType: String = "User"
+  given Encoder[User] = deriveWithDiscriminator
```

Comment:
do we now need this `...WithDiscriminator` version of encoder? I thought using the `AdtEncodingStrategy.flat` would do this automatically?

Reply:
Right, it would if we define only `given Encoder[Entity] = deriveAllEncoder[Entity]` but then we cannot define codecs on the leaf types like Project and User. It's an either-or situation. My thinking was that since it's nice to have these leaf entity codes available to be used on `getProject` like methods. I'm guessing, in a case we've got `Codec[Entity]` solely, decoding to `Entity` would be the only possibility followed by pattern match or `asInstanceOf`. Effectively, once you have `Codec[Entity]` only, there's no `Encoder[User]` available and `Encoder` is invariant.

Reply:
Ah right, thanks. Did not think about that case. But it's better!